txindex/include/persistor.h

```diff
@@ -0,0 +1,156 @@
+//
+// Created by os on 4/15/22.
+//
+#include <butil/macros.h>
+#include <gflags/gflags.h>
+#include "bthread/bthread.h"
+#include "bthread/mutex.h"
+#include <memory>
+#include <brpc/channel.h>
+#include "index.h"
+#include "service/storage/storage.pb.h"
+
+#ifndef AZINO_TXINDEX_INCLUDE_PERSISTOR_H
+#define AZINO_TXINDEX_INCLUDE_PERSISTOR_H
+DEFINE_int32(persist_period, 10000, "persist period time. measurement: millisecond");
+DEFINE_string(storage_addr, "0.0.0.0:8000", "Address of storage");
+namespace azino {
+    namespace txindex {
+
+        class Persistor {
+        public:
+            Persistor(TxIndex *index, uint32_t bucket_num) : _txindex(index), _bid(0),
+                                                             _bucket_num(bucket_num),
+                                                             _current_bucket(0) {
+
+
+                // Initialize the channel, NULL means using default options.
+
+                brpc::ChannelOptions option;
+                if (_channel.Init(FLAGS_storage_addr.c_str(), "", &option) != 0) {
+                    LOG(ERROR) << "Fail to initialize channel";
+                }
+                _stub.reset(new storage::StorageService_Stub(&_channel));
+            };
+
+
+            DISALLOW_COPY_AND_ASSIGN(Persistor);
+
+            ~Persistor() = default;
+
+            //Start a new thread and monitor the data. GetPersisting data periodically. Return 0 if success.
+            int Start() {
+                std::lock_guard<bthread::Mutex> lck(_mutex);
+                if (_bid != 0) {
+                    return -1;
+                }
+                return bthread_start_background(&_bid, &BTHREAD_ATTR_NORMAL, execute, this);
+            }
+
+            //Stop monitor thread. Need call first before the monitoring data destroy. Return 0 if success.
+            int Stop() {
+
+                bthread_t bid;
+
+                {// reset _bid, if the bthread wake up and found a different _bid, it will exit.
+                    std::lock_guard<bthread::Mutex> lck(_mutex);
+                    if (_bid == 0) {
+                        return -1;
+                    }
+                    bid = _bid;
+                    _bid = 0;
+                }
+
+
+                auto code = bthread_stop(bid);
+                return code;
```

Comment:
```suggestion
                {
                    // reset _bid, if the bthread wake up and found a different _bid, it will exit.
                    std::lock_guard<bthread::Mutex> lck(_mutex);
                    if (_bid == 0) {
                        return -1;
                    }
                    bid = _bid;
                    _bid = 0;
                    bthread_stop(bid);
                }

                return bthresd_join(bid, NULL);
```